<a href="https://colab.research.google.com/github/khing-fm/Data_Science/blob/main/pandas_for_data_analysi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import pandas Iconvention: import as 'pd')
import pandas as pd
import numpy as np

print(f"pandas version: {pd.__version__}")
print(f"numpy version: {np.__version__}")

pandas version: 2.2.2
numpy version: 2.0.2


In [ ]:
np.arange(0, 15, 2)

array([ 0,  2,  4,  6,  8, 10, 12, 14])

In [ ]:
# Example: Creating a Series

# From a list
sales = pd.Series([100, 150, 200, 175, 225])
print("series from list:")
print(sales)
print(sales.index)
print(f"\nIndex: {sales.index.tolist()}")
print(f"Values: {sales.values.tolist()}")

series from list:
0    100
1    150
2    200
3    175
4    225
dtype: int64
RangeIndex(start=0, stop=5, step=1)

Index: [0, 1, 2, 3, 4]
Values: [100, 150, 200, 175, 225]


In [ ]:
sales_mean = sales.mean()
sales_std = sales.std()

print(f"Mean Sales: {sales_mean}")
print(f"Standard Deviation of Sales: {sales_std}")

Mean Sales: 170.0
Standard Deviation of Sales: 48.088460154178364


### Sales Data Analysis Summary

*   **Mean Sales (170.0)**: Represents the average value of the sales entries.
*   **Standard Deviation (~48.09)**: Indicates the spread of the data; a higher value suggests that the sales figures vary significantly from the average.

In [ ]:
pd.Series(np.array([10, 20, 30, 40, 50]), dtype="int16")

,0
0,10
1,20
2,30
3,40
4,50


### Creating a Series with NumPy and dtypes

This example demonstrates:
1.  **NumPy Integration**: Creating a Series directly from an `np.array`.
2.  **Memory Optimization**: Using `dtype="int16"` to store integers using less memory (16 bits) compared to the standard 64-bit default.

In [ ]:
# Example: Series with custom index

sales = pd.Series(
    [100, 150, 200, 175, 225],
    index=["Jan", "Feb", "Mar", "Apr", "May"]
)
print("series with custom index:")
print(sales)
print(sales.index)
print(f"\nIndex: {sales.index.tolist()}")
print(f"Values: {sales.values.tolist()}")

# Acess by label
print(f"\nSales in May: {sales['May']}")
print(f"Jan to Mar: \n{sales['Jan':'Mar']}")

series with custom index:
Jan    100
Feb    150
Mar    200
Apr    175
May    225
dtype: int64
Index(['Jan', 'Feb', 'Mar', 'Apr', 'May'], dtype='object')

Index: ['Jan', 'Feb', 'Mar', 'Apr', 'May']
Values: [100, 150, 200, 175, 225]

Sales in May: 225
Jan to Mar: 
Jan    100
Feb    150
Mar    200
dtype: int64


### Custom Indexing and Label-Based Access

Pandas allows you to define a **custom index** (labels) for your data instead of just using default integers:

*   **Custom Labels**: In this example, we used month names (`Jan`, `Feb`, etc.) as the index.
*   **Access by Label**: You can retrieve specific values using the label name, like `sales['May']`.
*   **Slicing with Labels**: Unlike Python lists, when you slice with labels (`'Jan':'Mar'`), the **end point is included** in the result.

In [ ]:
employee = {
    "id": "101",
    "name": "Alex Johnson",
    "email": "alex.johnson@example-pet-store.com",
    "is_active": True,
    "department": "HR"
}

In [ ]:
import pandas as pd

pd.Series(employee)

,0
id,101
name,Alex Johnson
email,alex.johnson@example-pet-store.com
is_active,True
department,HR


In [ ]:
staff_data = {
    "staff1":{
        "id": 101,
        "name": "John Doe",
        "email": "john.mclean@examplepetstore.com",
        "is_active": True,
        "department": "HR"
    },
    "staff2":{
        "id": 102,
        "name": "Jane Smith",
        "email": "john.catron@example-pet-store.com",
        "is_active": False,
        "department": "IT"
    },
    "staff3":{
        "id": 103,
        "name": "Bob Johnson",
        "email": "william.paterson@my-own-personal-domain.com",
        "is_active": True,
        "department": "Finance"
    },
    "staff4":{
        "id": 104,
        "name": "Alice Brown",
        "email": "john.mclean@examplepetstore.com",
        "is_active": True,
        "department": "Marketing"
    },
}

In [ ]:
pd.DataFrame(staff_data)

,staff1,staff2,staff3,staff4
id,101,102,103,104
name,John Doe,Jane Smith,Bob Johnson,Alice Brown
email,john.mclean@examplepetstore.com,john.catron@example-pet-store.com,william.paterson@my-own-personal-domain.com,john.mclean@examplepetstore.com
is_active,True,False,True,True
department,HR,IT,Finance,Marketing


In [ ]:
employee = {
    "name": ["John Doe", "Jane Smith", "Bob Johnson", "Alice Brown"],
    "id": [101, 102, 103, 104],
    "email": [
        "john.doe@example.com",
        "jane.smith@example.com",
        "bob.johnson@example.com",
        "alice.brown@example.com"
    ],
    "is_active": [True, False, True, True],
    "department": [["HR", "Finance"], ["IT"], ["Marketing"], ["HR", "Finance", "Marketing"]
    ]
}
# Dictionary of lists

In [ ]:
pd.DataFrame(employee)

,name,id,email,is_active,department
0,John Doe,101,john.doe@example.com,True,"[HR, Finance]"
1,Jane Smith,102,jane.smith@example.com,False,[IT]
2,Bob Johnson,103,bob.johnson@example.com,True,[Marketing]
3,Alice Brown,104,alice.brown@example.com,True,"[HR, Finance, Marketing]"


In [ ]:
df = pd.DataFrame(employee)
df.info()
df.shape
df.ndim
df.size
df.columns
df.index
df.values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        4 non-null      object
 1   id          4 non-null      int64 
 2   email       4 non-null      object
 3   is_active   4 non-null      bool  
 4   department  4 non-null      object
dtypes: bool(1), int64(1), object(3)
memory usage: 264.0+ bytes


array([['John Doe', 101, 'john.doe@example.com', True,
        list(['HR', 'Finance'])],
       ['Jane Smith', 102, 'jane.smith@example.com', False, list(['IT'])],
       ['Bob Johnson', 103, 'bob.johnson@example.com', True,
        list(['Marketing'])],
       ['Alice Brown', 104, 'alice.brown@example.com', True,
        list(['HR', 'Finance', 'Marketing'])]], dtype=object)

In [ ]:
df

,name,id,email,is_active,department
0,John Doe,101,john.doe@example.com,True,"[HR, Finance]"
1,Jane Smith,102,jane.smith@example.com,False,[IT]
2,Bob Johnson,103,bob.johnson@example.com,True,[Marketing]
3,Alice Brown,104,alice.brown@example.com,True,"[HR, Finance, Marketing]"


In [ ]:
df [["name", "email", "department"]]

,name,email,department
0,John Doe,john.doe@example.com,"[HR, Finance]"
1,Jane Smith,jane.smith@example.com,[IT]
2,Bob Johnson,bob.johnson@example.com,[Marketing]
3,Alice Brown,alice.brown@example.com,"[HR, Finance, Marketing]"
